// xiome/x/toolbox/dbby/dbby-constrain.js
import { objectMap } from "../object-map.js";
import { and } from "./dbby-helpers.js";
export function dbbyConstrain(table, namespace) {
    const spike = (conditionTree) => (conditionTree
        ? and({ equal: namespace }, conditionTree)
        : and({ equal: namespace }));
    return {
        async create(...rows) {
            return table.create(...rows.map(row => ({ ...row, ...namespace })));
        },
        async read(options) {
            return table.read({
                ...options,
                conditions: spike(options.conditions),
            });
        },
        async one(options) {
            return table.one({
                ...options,
                conditions: spike(options.conditions),
            });
        },
        async assert(options) {
            return table.assert({
                ...options,
                conditions: spike(options.conditions),
                make: async () => {
                    const row = await options.make();
                    return { ...row, ...namespace };
                },
            });
        },
        async update(options) {
            const opts = options;
            return table.update({
                ...options,
                conditions: spike(options.conditions),
                upsert: opts.upsert
                    ? { ...opts.upsert, ...namespace }
                    : undefined,
                whole: opts.whole
                    ? { ...opts.whole, ...namespace }
                    : undefined,
                write: opts.write
                    ? { ...opts.write, ...namespace }
                    : undefined,
            });
        },
        async delete(options) {
            return table.delete({
                ...options,
                conditions: spike(options.conditions),
            });
        },
        async count(options) {
            return table.count({
                ...options,
                conditions: spike(options.conditions),
            });
        },
    };
}
export function prepareConstrainTables(tables) {
    return (namespace) => objectMap(tables, table => dbbyConstrain(table, namespace));
}
