// xiome/x/features/auth/models/personal-model.js
import { ops } from "../../../framework/ops.js";
import { autowatcher } from "../../../toolbox/autowatcher/autowatcher.js";
export function makePersonalModel({ personalService, getAccess, reauthorize, }) {
    const auto = autowatcher();
    const state = auto.state({
        submitDraftOp: ops.ready(undefined)
    });
    const actions = auto.actions({
        setSubmitDraft(op) {
            state.submitDraftOp = op;
        },
    });
    return {
        get submitDraftOp() {
            return state.submitDraftOp;
        },
        async saveProfile(profileDraft) {
            await ops.operation({
                promise: (async () => {
                    const { user: { userId } } = await getAccess();
                    await personalService.setProfile({ userId, profileDraft });
                    await reauthorize();
                })(),
                errorReason: "error saving profile",
                setOp(op) {
                    actions.setSubmitDraft(op);
                },
            });
        }
    };
}
