// xiome/x/features/store/api/store-api.js
import { apiContext } from "renraku/x/api/api-context.js";
import { shoppingTopic } from "../topics/shopping-topic.js";
import { shopkeepingTopic } from "../topics/shopkeeping-topic.js";
import { stripeConnectTopic } from "../topics/stripe-connect-topic.js";
import { payPolicies } from "./policies/store-policies.js";
import { statusCheckerTopic } from "../topics/status-checker-topic.js";
import { statusTogglerTopic } from "../topics/status-toggler-topic.js";
export const storeApi = ({ rando, tables, config, authPolicies, stripeComplex, shoppingOptions, stripeConnectOptions, }) => {
    const policies = payPolicies({
        tables,
        authPolicies,
        stripeComplex,
    });
    return {
        stripeConnectService: apiContext()({
            policy: policies.merchant,
            expose: stripeConnectTopic(stripeConnectOptions),
        }),
        shopkeepingService: apiContext()({
            policy: policies.clerk,
            expose: shopkeepingTopic({
                generateId: () => rando.randomId(),
            }),
        }),
        shoppingService: apiContext()({
            policy: policies.customer,
            expose: shoppingTopic(shoppingOptions),
        }),
        ecommerce: {
            statusTogglerService: apiContext()({
                policy: policies.clerk,
                expose: statusTogglerTopic(),
            }),
            statusCheckerService: apiContext()({
                policy: policies.prospect,
                expose: statusCheckerTopic({ config }),
            }),
        }
    };
};
