// renraku/x/remote/make-json-requester.js
import { capitalizedHeaders } from "../tools/capitalized-headers.js";
import { jsonHttpRequest } from "../jsonrpc/json-http-request.js";
import { parseJsonResponse } from "../jsonrpc/parse-json-response.js";
export function makeJsonRequester({ fetch, headers }) {
    return async function requester({ link, args, meta, specifier }) {
        const request = jsonHttpRequest({
            link,
            args,
            meta,
            headers,
            specifier,
        });
        const fetchRequest = new window.Request(link, {
            body: request.body,
            method: request.method,
            headers: capitalizedHeaders(request.headers),
        });
        const fetchResponse = await fetch(fetchRequest);
        const collectHeaders = {};
        fetchResponse.headers.forEach((value, key) => collectHeaders[key] = value);
        const responseHeaders = collectHeaders;
        const result = parseJsonResponse({
            headers: responseHeaders,
            status: fetchResponse.status,
            body: await fetchResponse.text(),
        });
        return result;
    };
}
