// xiome/x/features/auth/topics/app-edit-topic.js
import { asTopic } from "renraku/x/identities/as-topic.js";
import { throwProblems } from "../../../toolbox/topic-validation/throw-problems.js";
import { find } from "../../../toolbox/dbby/dbby-mongo.js";
import { validateAppDraft } from "./apps/validate-app-draft.js";
import { originsToDatabase } from "./origins/origins-to-database.js";
export const appEditTopic = (options) => asTopic()({
    async updateApp({ tables, access }, { appId, appDraft }) {
        // await requireUserIsAllowedToEditApp({tables, access, appId})
        throwProblems(validateAppDraft(appDraft));
        await tables.app.app.update({
            ...find({ appId }),
            whole: {
                appId,
                home: appDraft.home,
                label: appDraft.label,
                origins: originsToDatabase(appDraft.origins),
                archived: false,
            },
        });
    },
    async deleteApp({ tables, access }, { appId }) {
        // await requireUserIsAllowedToEditApp({tables, access, appId})
        await tables.app.app.update({
            ...find({ appId }),
            write: { archived: true },
        });
    },
});
