// xiome/x/assembly/frontend/connect/mock/common/mock-register-app.js
export async function mockRegisterApp({ apiLink, ownerEmail, appOrigins, backend }) {
    const mockBrowserForPlatform = await backend.mockBrowser();
    const mockWindowForPlatform = await mockBrowserForPlatform.mockAppWindow({
        apiLink,
        latency: false,
        appId: backend.platformAppId,
        windowLink: window.location.href,
    });
    const { authModel, appModel } = mockWindowForPlatform.models;
    await authModel.sendLoginLink(ownerEmail);
    await authModel.login(backend.emails.recallLatestLoginEmail().loginToken);
    const { appId } = await appModel.registerApp({
        label: "Mock App",
        home: window.location.href,
        origins: appOrigins,
    });
    console.log("mock: register app", appId);
    // TODO reactivate store
    // // link bank account with stripe
    // await mockWindowForPlatform.models.storeModel.shares.bank.setupStripeAccount(appId)
    // const stripeDetails = await mockWindowForPlatform.models.bankModel.getStripeAccountDetails(appId)
    // console.log("mock: app stripe details", stripeDetails)
    return appId;
}
